Rubberduck.CodeAnalysis/Inspections/Concrete/PublicImplementationShouldBePrivateInspection.cs

```diff
@@ -0,0 +1,95 @@
+﻿using Rubberduck.CodeAnalysis.CodeMetrics;
+using Rubberduck.CodeAnalysis.Inspections.Abstract;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Parsing.VBA.DeclarationCaching;
+using Rubberduck.Refactorings.Common;
+using Rubberduck.Resources.Inspections;
+using Rubberduck.VBEditor;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Rubberduck.CodeAnalysis.Inspections.Concrete
+{
+    internal sealed class PublicImplementationShouldBePrivateInspection : DeclarationInspectionBase
+    {
+        public PublicImplementationShouldBePrivateInspection(IDeclarationFinderProvider declarationFinderProvider)
+            : base(declarationFinderProvider, DeclarationType.Member)
+        {}
+
+        //Overriding DoGetInspectionResults in order to dereference the DeclarationFinder FindXXX declaration 
+        //lists only once per inspections pass.
+        protected override IEnumerable<IInspectionResult> DoGetInspectionResults(DeclarationFinder finder)
+        {
+            var publicMembers = finder.UserDeclarations(DeclarationType.Member)
+                .Where(d => !d.HasPrivateAccessibility()
+                    && IsLikeAnImplementerOrHandlerName(d.IdentifierName));
+
+            if (!publicMembers.Any())
+            {
+                return Enumerable.Empty<IInspectionResult>();
+            }
+
+            var publicImplementersAndHandlers = finder.FindAllInterfaceImplementingMembers()
+                .Where(d => !d.HasPrivateAccessibility())
+                .Concat(finder.FindEventHandlers()
+                    .Where(d => !d.HasPrivateAccessibility()));
+
+            var publicDocumentEvents = FindDocumentEventHandlers(publicMembers);
+
+            return publicMembers.Intersect(publicImplementersAndHandlers)
+                .Concat(publicDocumentEvents)
+                .Select(InspectionResult)
+                .ToList();
+        }
+
+        private static IEnumerable<Declaration> FindDocumentEventHandlers(IEnumerable<Declaration> publicMembers)
```

Comment:
I was/am unable to find any `DeclarationFinder` support for document event handlers.  Hence the, admittedly weaker, name-based approach.  Am I looking in the right class?

Reply:
Ah that's too bad, I recall there was work done to get document modules' actual class type so we could tell Worksheet apart from Workbook; I thought we had followed through with resolving their respective events... that said maybe we didn't, in which case a purely name-based approach makes sense.
That said since we do have the proper interfaces and class types for these modules (seems all we got is the identifier name, under `DocumentModuleDeclaration.SupertypeNames`), I wonder if we couldn't get the document-module event names from the declaration finder. It's possible that there's some wiggling with an underscore prefix to get the correct type; if that works, then we can probably tweak the finder to make these events easier to lookup in places such as inspections.
That said I think the current approach is probably fine for an initial release, not a showstopper at all!

Reply:
I looked too quickly, the declarations for each supertypes are indeed resolved. If I can get my local build to stop pointing to RD3 I'll have a PR for pulling the document events into the finder later, assuming it works.

Reply:
Sounds awesome!  I'm sure having them available from the finder will have plenty of uses.